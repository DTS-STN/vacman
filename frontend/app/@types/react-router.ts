import 'react-router';

import type { Namespace } from 'i18next';

declare module 'react-router' {
  interface AppLoadContext {
    nonce: string;
    session: AppSession;
  }

  /**
   * Route handles should export an i18n namespace, if necessary.
   */
  interface RouteHandle {
    i18nNamespace?: Namespace;
  }

  /**
   * A route module exports an optional RouteHandle.
   */
  interface RouteModule {
    handle?: RouteHandle;
  }

  /**
   * Override the default React Router RouteModules
   * to include the new RouteModule type.
   */
  interface RouteModules extends Record<string, RouteModule | undefined> {}
}

export {};
