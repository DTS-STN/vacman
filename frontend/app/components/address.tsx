import type { ComponentProps, JSX } from 'react';

import { formatAddress } from '~/utils/string-utils';
import { cn } from '~/utils/tailwind-utils';

type AddressProps = ComponentProps<'address'> & {
  address: Parameters<typeof formatAddress>[0];
  format?: Parameters<typeof formatAddress>[1];
};

export function Address({ address, format, className, ...rest }: AddressProps): JSX.Element {
  return (
    <address className={cn('whitespace-pre-wrap not-italic', className)} data-testid="address-id" {...rest}>
      {formatAddress(address, format)}
    </address>
  );
}
