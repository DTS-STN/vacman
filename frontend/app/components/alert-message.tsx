import type { Ref } from 'react';

import type { IconDefinition } from '@fortawesome/free-solid-svg-icons';
import { faTriangleExclamation, faCircleExclamation, faCircleCheck } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

interface AlertMessageProps {
  message: string;
  type: 'success' | 'info' | 'error';
  ref?: Ref<HTMLDivElement>;
}

const styles: Record<NonNullable<AlertMessageProps['type']>, string> = {
  success: `border-green-600 bg-green-100`,
  error: `border-red-800 bg-red-100`,
  info: `border-sky-700 bg-sky-100`,
};

const icons: Record<NonNullable<AlertMessageProps['type']>, IconDefinition> = {
  success: faCircleCheck,
  error: faTriangleExclamation,
  info: faCircleExclamation,
};

export function AlertMessage({ ref, message, type = 'info' }: AlertMessageProps) {
  return (
    <div ref={ref} className={`${styles[type]} flex w-full items-center border-l-4 p-2`}>
      <FontAwesomeIcon
        icon={icons[type]}
        className={`${type === 'success' ? 'text-green-600' : type === 'error' ? 'text-red-800' : 'text-sky-700'} mr-2`}
      />
      {message}
    </div>
  );
}
