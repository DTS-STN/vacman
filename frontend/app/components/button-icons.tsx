import type { FontAwesomeIconProps } from '@fortawesome/react-fontawesome';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import { cn } from '~/utils/tailwind-utils';

export function ButtonStartIcon({ className, ...restProps }: FontAwesomeIconProps) {
  return <FontAwesomeIcon fixedWidth className={cn('me-1', className)} {...restProps} />;
}

export function ButtonEndIcon({ className, ...restProps }: FontAwesomeIconProps) {
  return <FontAwesomeIcon fixedWidth className={cn('ms-1', className)} {...restProps} />;
}
