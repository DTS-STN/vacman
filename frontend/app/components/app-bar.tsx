import type { JSX } from 'react';

import { faChevronDown, faRightFromBracket, faUser } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useTranslation } from 'react-i18next';

import { DropdownMenu, DropdownMenuContent, DropdownMenuLabel, DropdownMenuTrigger } from '~/components/dropdown-menu';
import { LanguageSwitcher } from '~/components/language-switcher';
import { Menu, MenuItem } from '~/components/menu';
import { cn } from '~/utils/tailwind-utils';

type AppBarProps = {
  children: React.ReactNode;
  name?: string;
  profileItems?: React.ReactNode;
};

export function AppBar({ children, name, profileItems }: AppBarProps): JSX.Element {
  const { t } = useTranslation(['gcweb']);

  return (
    <div className="bg-slate-700">
      <div className="align-center container mx-auto flex flex-wrap justify-between">
        <div className="align-center flex">
          <Menu>{children}</Menu>
        </div>
        <div className="flex items-center space-x-4 text-right">
          {name && <UserButton name={name}>{profileItems}</UserButton>}
          <LanguageSwitcher>{t('gcweb:language-switcher.alt-lang')}</LanguageSwitcher>
        </div>
      </div>
    </div>
  );
}

type UserButtonProps = {
  className?: string;
  children?: React.ReactNode;
  name?: string;
};

function UserButton({ className, children, name }: UserButtonProps): JSX.Element {
  const { t } = useTranslation(['gcweb']);

  return (
    <DropdownMenu>
      <DropdownMenuTrigger
        className={cn(
          'flex h-full flex-nowrap space-x-2 bg-slate-600 px-2 text-sm text-white hover:bg-slate-500 hover:underline focus:ring-2 focus:ring-black focus:ring-offset-2 focus:outline-hidden aria-expanded:bg-slate-800 aria-expanded:text-white sm:space-x-4 sm:px-4',
          className,
        )}
      >
        <div className="text-md my-auto flex flex-nowrap items-center space-x-2 py-2">
          <div className="flex h-8 w-8 items-center justify-center rounded-full bg-white">
            <FontAwesomeIcon icon={faUser} className="size-5 text-slate-700" />
          </div>
          <span id="menu-label" className="text-md hidden py-2 font-bold sm:block">
            {name}
          </span>
        </div>
        <FontAwesomeIcon icon={faChevronDown} className="my-auto size-4" />
      </DropdownMenuTrigger>
      <DropdownMenuContent align="end" className="w-64 bg-slate-700">
        <UserName name={name} />
        <MenuItem file="routes/protected/index.tsx">{t('gcweb:app.profile')}</MenuItem>
        {children}
        <MenuItem to="/auth/logout" className="text-md flex justify-between">
          {t('gcweb:app.logout')}
          <FontAwesomeIcon icon={faRightFromBracket} className="my-auto size-8" />
        </MenuItem>
      </DropdownMenuContent>
    </DropdownMenu>
  );
}

type UserNameProps = {
  name?: string;
};

function UserName({ name }: UserNameProps): JSX.Element {
  return (
    <>
      {name !== undefined && (
        <DropdownMenuLabel className="text-md flex items-center border-b-2 border-slate-600 px-3 py-2 text-gray-300 sm:hidden">
          <FontAwesomeIcon icon={faUser} className="mr-2 size-4" />
          {name}
        </DropdownMenuLabel>
      )}
    </>
  );
}
