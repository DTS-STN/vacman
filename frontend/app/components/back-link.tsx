import type { Params, Path } from 'react-router';

import { faAngleLeft } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { useTranslation } from 'react-i18next';

import { InlineLink } from '~/components/links';
import type { I18nRouteFile } from '~/i18n-routes';
import { cn } from '~/utils/tailwind-utils';

interface BackLinkProps {
  file: I18nRouteFile;
  params?: Params;
  className?: string;
  hash?: Path['hash'];
  search?: Path['search'];
  translationKey?: string;
}

export function BackLink({ className, file, params, hash, search, translationKey = 'app:profile.back' }: BackLinkProps) {
  const { t } = useTranslation();
  const linkText = t(translationKey);
  return (
    <InlineLink
      className={cn('inline-flex items-center', className)}
      file={file}
      params={params}
      hash={hash}
      search={search}
      aria-label={linkText}
    >
      <FontAwesomeIcon icon={faAngleLeft} />
      {linkText}
    </InlineLink>
  );
}
