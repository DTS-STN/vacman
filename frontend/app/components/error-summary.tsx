import type { ComponentPropsWithoutRef, JSX } from 'react';
import { useEffect, useRef, useState } from 'react';

import { useActionData, useFetcher } from 'react-router';

import { useTranslation } from 'react-i18next';

import { AnchorLink } from '~/components/links';
import { cn } from '~/utils/tailwind-utils';

/**
 * Represents an item in the error summary, containing an error message
 * and the associated field ID where the error occurred.
 */
export interface ErrorSummaryItem {
  /** The error message to display. */
  errorMessage: string;
  /** The ID of the field associated with the error. */
  fieldId: string;
}

/**
 * Props for the ErrorSummary component.
 */
export interface ErrorSummaryProps extends OmitStrict<ComponentPropsWithoutRef<'section'>, 'role' | 'tabIndex'> {
  /** A list of error items to display. */
  errors: readonly ErrorSummaryItem[];
}

/**
 * ErrorSummary component displays a list of error messages with associated
 * field IDs and provides anchor links for users to quickly navigate to
 * specific form fields that caused the errors.
 *
 * - When there are errors, the component scrolls the section into view
 *   and focuses on it to ensure visibility.
 * - Provides an accessible, styled list of errors with links to the form fields.
 *
 * @returns The rendered component or `null` if there are no errors.
 */
export function ErrorSummary({ className, errors, ...rest }: ErrorSummaryProps): JSX.Element | null {
  const { t } = useTranslation(['gcweb']);
  const sectionRef = useRef<HTMLElement>(null);

  // Scroll and focus on the error summary when errors are updated.
  useEffect(() => {
    if (errors.length > 0 && sectionRef.current) {
      sectionRef.current.scrollIntoView({ behavior: 'smooth' });
      sectionRef.current.focus();
    }
  }, [errors]);

  if (errors.length === 0) {
    return null;
  }

  return (
    <section
      ref={sectionRef}
      className={cn('my-5 border-4 border-red-600 p-4', className)}
      {...rest}
      tabIndex={-1}
      role="alert"
    >
      <h2 className="font-lato text-lg font-semibold">{t('gcweb:error-summary.header', { count: errors.length })}</h2>
      <ol className="mt-1.5 list-decimal space-y-2 pl-7">
        {errors.map(({ fieldId, errorMessage }) => (
          <li key={`${fieldId}-${errorMessage}`}>
            <AnchorLink className="text-red-700 underline hover:decoration-2 focus:decoration-2" anchorElementId={fieldId}>
              {errorMessage}
            </AnchorLink>
          </li>
        ))}
      </ol>
    </section>
  );
}

/**
 * Props for the `ActionDataErrorSummary` component.
 */
type ActionDataErrorSummaryProps = ComponentPropsWithoutRef<'div'> & {
  /**
   * The action data from a React Router Form or Fetcher.
   * This triggers error collection when form or Fetcher data changes.
   */
  actionData: unknown | undefined;
};

/**
 * The `ActionDataErrorSummary` component collects and displays form validation errors.
 *
 * @remarks
 * This component uses the `useEffect` hook to capture validation errors from form inputs that contain
 * the `aria-errormessage` attribute. It finds the error messages, associates them with input fields,
 * and displays them using the `ErrorSummary` component.
 *
 * @example
 * ```tsx
 * import { Form } from 'react-router';
 * import type { Route } from './+types/my-route';
 * import { ActionDataErrorSummary } from '~/components/error-summary';
 *
 * export default function MyRouteComponent({ loaderData, actionData, params }: Route.ComponentProps) {
 *   return (
 *     <ActionDataErrorSummary actionData={actionData}>
 *       <Form method="post" noValidate>
 *         ...
 *       </Form>
 *     </ActionDataErrorSummary>
 *   );
 * }
 * ```
 */
export function ActionDataErrorSummary({ actionData, children, ...rest }: ActionDataErrorSummaryProps): JSX.Element {
  const [errors, setErrors] = useState<readonly ErrorSummaryItem[]>([]);
  const containerRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (!containerRef.current) return;
    setErrors(collectFormValidationErrors(containerRef.current));
  }, [actionData]);

  return (
    <div ref={containerRef} {...rest}>
      {errors.length > 0 && <ErrorSummary errors={errors} />}
      {children}
    </div>
  );
}

/**
 * Collects validation errors from form inputs within a container element.
 * @param container - The DOM element to search for form inputs with validation errors
 * @returns An array of ErrorSummaryItems containing field IDs and their corresponding error messages
 */
function collectFormValidationErrors(container: HTMLElement): readonly ErrorSummaryItem[] {
  // Keeps track of processed error message IDs to prevent duplicate error messages
  // when multiple inputs reference the same error message element
  const processedErrorMessageIds = new Set<string>();

  const errorItems: ErrorSummaryItem[] = [];

  for (const input of findInvalidInputs(container)) {
    for (const messageId of getErrorMessageIds(input)) {
      // Skip this error message if we've already processed it to avoid duplicates
      // This can happen when multiple form fields reference the same error message element
      if (processedErrorMessageIds.has(messageId)) {
        continue;
      }

      const errorItem = createErrorItem(input, messageId);

      if (errorItem) {
        errorItems.push(errorItem);
        processedErrorMessageIds.add(messageId);
      }
    }
  }

  return errorItems;
}

/**
 * Finds all input elements with aria-errormessage attributes in the container
 */
function findInvalidInputs(container: HTMLElement): readonly Element[] {
  return Array.from(container.querySelectorAll('[aria-errormessage]'));
}

/**
 * Extracts and returns array of error message IDs from an input's aria-errormessage attribute
 */
function getErrorMessageIds(input: Element): readonly string[] {
  const errorMessageAttr = input.getAttribute('aria-errormessage');

  if (!errorMessageAttr) {
    return [];
  }

  return errorMessageAttr.split(' ').filter(Boolean);
}

/**
 * Creates an ErrorSummaryItem from an input element and error message ID
 * @returns ErrorSummaryItem if valid error message is found, null otherwise
 */
function createErrorItem(input: Element, errorMessageId: string): ErrorSummaryItem | null {
  const errorMessageElement = document.getElementById(errorMessageId);
  const errorMessageContent = errorMessageElement?.textContent?.trim();

  if (!errorMessageContent) {
    return null;
  }

  return {
    fieldId: input.id,
    errorMessage: errorMessageContent,
  };
}

/**
 * Props for the `FetcherErrorSummary` component.
 */
type FetcherErrorSummaryProps = OmitStrict<ActionDataErrorSummaryProps, 'actionData'> & {
  /**
   * The key of the Fetcher instance to use for error collection.
   */
  fetcherKey: string;
};

/**
 * `FetcherErrorSummary` component collects and displays validation errors from a specific Fetcher.
 *
 * @remarks
 * Fetchers are instances provided by React Router that manage form interactions independently.
 * This component works by linking to a specific fetcher instance identified by the `fetcherKey`.
 *
 * @example
 * ```tsx
 * import { useFetcher } from 'react-router';
 * import type { Route } from './+types/my-route';
 * import { FetcherErrorSummary } from '~/components/error-summary';
 *
 * export default function MyRouteComponent({ loaderData, actionData, params }: Route.ComponentProps) {
 *   const fetcher = useFetcher({ key: "my-fetcher-key" });
 *
 *   return (
 *     <FetcherErrorSummary fetcherKey="my-fetcher-key">
 *       <fetcher.Form method="post" noValidate>
 *         ...
 *       </fetcher.Form>
 *     </FetcherErrorSummary>
 *   );
 * }
 * ```
 */
export function FetcherErrorSummary({ children, fetcherKey, ...rest }: FetcherErrorSummaryProps): JSX.Element {
  const fetcher = useFetcher({ key: fetcherKey });
  return (
    <ActionDataErrorSummary actionData={fetcher.data} {...rest}>
      {children}
    </ActionDataErrorSummary>
  );
}

/**
 * Props for the `FormErrorSummary` component.
 */
type FormDataErrorSummaryProps = OmitStrict<ActionDataErrorSummaryProps, 'actionData'>;

/**
 * `FormErrorSummary` component collects and displays validation errors from a form.
 *
 * @remarks
 * This component automatically links to the action data of the current form by using the `useActionData` hook.
 *
 * @example
 * ```tsx
 * import { Form } from 'react-router';
 * import type { Route } from './+types/my-route';
 * import { FormErrorSummary } from '~/components/error-summary';
 *
 * export default function MyRouteComponent({ loaderData, actionData, params }: Route.ComponentProps) {
 *   return (
 *     <FormErrorSummary>
 *       <Form method="post" noValidate>
 *         ...
 *       </Form>
 *     </FormErrorSummary>
 *   );
 * }
 * ```
 */
export function FormErrorSummary({ children, ...rest }: FormDataErrorSummaryProps): JSX.Element {
  const actionData = useActionData();
  return (
    <ActionDataErrorSummary actionData={actionData} {...rest}>
      {children}
    </ActionDataErrorSummary>
  );
}
