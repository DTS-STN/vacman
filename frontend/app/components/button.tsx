import type { ComponentProps } from 'react';

import { ButtonEndIcon, ButtonStartIcon } from '~/components/button-icons';
import { cn } from '~/utils/tailwind-utils';

type ButtonEndIconProps = ComponentProps<typeof ButtonEndIcon>;
type ButtonStartIconProps = ComponentProps<typeof ButtonStartIcon>;

const sizes = {
  xs: 'px-3 py-2 text-xs',
  sm: 'px-3 py-2 text-sm',
  base: 'px-5 py-2.5 text-sm',
  lg: 'px-5 py-3 text-base',
  xl: 'px-6 py-3.5 text-base',
} as const;

// prettier-ignore
const variants = {
  alternative: 'border-gray-200 bg-white text-gray-900 hover:bg-gray-100 hover:text-blue-700 focus:bg-gray-100 focus:text-blue-700',
  default: 'border-gray-300 bg-gray-200 text-slate-700 hover:bg-neutral-300 focus:bg-neutral-300',
  dark: 'border-gray-800 bg-gray-800 text-white hover:bg-gray-900 focus:bg-gray-900',
  green: 'border-green-700 bg-green-700 text-white hover:bg-green-800 focus:bg-green-800',
  primary: 'border-slate-700 bg-slate-700 text-white hover:bg-sky-800 focus:bg-sky-800',
  red: 'border-red-700 bg-red-700 text-white hover:bg-red-800 focus:bg-red-800',
  link: 'border-0 bg-white text-gray-900 underline hover:bg-gray-100 hover:text-blue-700 focus:bg-gray-100 focus:text-blue-700',
  ghost: 'border-0 hover:bg-neutral-100 hover:text-neutral-900',
  outline: "border border-neutral-200 bg-white hover:bg-neutral-100 hover:text-neutral-900 disabled:text-neutral-500",
} as const;

export interface ButtonProps extends ComponentProps<'button'> {
  endIcon?: ButtonEndIconProps['icon'];
  endIconProps?: OmitStrict<ButtonEndIconProps, 'icon'>;
  pill?: boolean;
  size?: keyof typeof sizes;
  startIcon?: ButtonStartIconProps['icon'];
  startIconProps?: OmitStrict<ButtonStartIconProps, 'icon'>;
  variant?: keyof typeof variants;
}

/**
 * Tailwind CSS Buttons from Flowbite
 * @see https://flowbite.com/docs/components/buttons/
 */
export function Button({
  children,
  className,
  endIcon,
  endIconProps,
  pill,
  size = 'base',
  startIcon,
  startIconProps,
  variant = 'default',
  ...props
}: ButtonProps) {
  const buttonClassName = cn(
    'font-lato inline-flex items-center justify-center rounded border align-middle outline-offset-4',
    sizes[size],
    variants[variant],
    'disabled:pointer-events-none disabled:cursor-not-allowed disabled:opacity-70',
    pill && 'rounded-full',
    className,
  );
  return (
    <button className={buttonClassName} {...props}>
      {startIcon && <ButtonStartIcon {...(startIconProps ?? {})} icon={startIcon} />}
      {children}
      {endIcon && <ButtonEndIcon {...(endIconProps ?? {})} icon={endIcon} />}
    </button>
  );
}
