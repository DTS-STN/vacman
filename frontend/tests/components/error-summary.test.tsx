import { useActionData, useFetcher } from 'react-router';

import { render } from '@testing-library/react';
import { describe, expect, it, vi } from 'vitest';
import { mock } from 'vitest-mock-extended';

import { ActionDataErrorSummary, ErrorSummary, FetcherErrorSummary, FormErrorSummary } from '~/components/error-summary';

// See https://github.com/jsdom/jsdom/issues/1695
Element.prototype.scrollIntoView = vi.fn();

vi.mock('react-router', () => ({
  useFetcher: vi.fn(),
  useActionData: vi.fn(),
}));

describe('ErrorSummary', () => {
  it('should not render an error summary when there are no errors', () => {
    const { container } = render(<ErrorSummary errors={[]} />);
    expect(container).toMatchSnapshot('expected html');
    expect(container.ownerDocument.activeElement !== container.querySelector('section')).toBeTruthy();
  });

  it('should render an error summary with 1 error', () => {
    const { container } = render(<ErrorSummary errors={[{ fieldId: 'input-test', errorMessage: 'Test error 1' }]} />);
    expect(container).toMatchSnapshot('expected html');
    expect(container.ownerDocument.activeElement === container.querySelector('section')).toBeTruthy();
  });

  it('should render an error summary with 2 errors', () => {
    const { container } = render(
      <ErrorSummary
        errors={[
          { fieldId: 'input-test', errorMessage: 'Test error 1' },
          { fieldId: 'input-test', errorMessage: 'Test error 2' },
        ]}
      />,
    );
    expect(container).toMatchSnapshot('expected html');
    expect(container.ownerDocument.activeElement === container.querySelector('section')).toBeTruthy();
  });
});

describe('ActionDataErrorSummary', () => {
  it('displays errors from action data', () => {
    const actionData = { errors: { field: ['Test error message'] } };

    const { container } = render(
      <ActionDataErrorSummary actionData={actionData}>
        <div id="test-field" aria-errormessage="test-error">
          Test Field
        </div>
        <div id="test-error">{actionData.errors.field.at(0)}</div>
      </ActionDataErrorSummary>,
    );

    expect(container).toMatchSnapshot('expected html');
    expect(container.ownerDocument.activeElement === container.querySelector('section')).toBeTruthy();
  });
});

describe('FetcherErrorSummary', () => {
  it('displays errors from fetcher data', () => {
    const actionData = { errors: { field: ['Test error message'] } };
    const fetcherMock = mock<ReturnType<typeof useFetcher>>({ data: actionData });
    vi.mocked(useFetcher).mockReturnValue(fetcherMock);

    const { container } = render(
      <FetcherErrorSummary fetcherKey="test-fetcher">
        <div id="test-field" aria-errormessage="test-error">
          Test Field
        </div>
        <div id="test-error">{actionData.errors.field.at(0)}</div>
      </FetcherErrorSummary>,
    );

    expect(container).toMatchSnapshot('expected html');
    expect(container.ownerDocument.activeElement === container.querySelector('section')).toBeTruthy();
  });
});

describe('FormErrorSummary', () => {
  it('displays errors from form action data', () => {
    const actionData = { errors: { field: ['Test error message'] } };
    vi.mocked(useActionData).mockReturnValue(actionData);

    const { container } = render(
      <FormErrorSummary>
        <div id="test-field" aria-errormessage="test-error">
          Test Field
        </div>
        <div id="test-error">{actionData.errors.field.at(0)}</div>
      </FormErrorSummary>,
    );

    expect(container).toMatchSnapshot('expected html');
    expect(container.ownerDocument.activeElement === container.querySelector('section')).toBeTruthy();
  });
});
