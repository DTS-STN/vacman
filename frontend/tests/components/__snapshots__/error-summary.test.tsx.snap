// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActionDataErrorSummary > displays errors from action data > expected html 1`] = `
<div>
  <div>
    <section
      class="my-5 border-4 border-red-600 p-4"
      role="alert"
      tabindex="-1"
    >
      <h2
        class="font-lato text-lg font-semibold"
      >
        {"key":"gcweb:error-summary.header","options":{"count":1}}
      </h2>
      <ol
        class="mt-1.5 list-decimal space-y-2 pl-7"
      >
        <li>
          <a
            class="text-red-700 underline hover:decoration-2 focus:decoration-2"
            data-testid="anchor-link"
            href="#test-field"
          >
            Test error message
          </a>
        </li>
      </ol>
    </section>
    <div
      aria-errormessage="test-error"
      id="test-field"
    >
      Test Field
    </div>
    <div
      id="test-error"
    >
      Test error message
    </div>
  </div>
</div>
`;

exports[`ErrorSummary > should not render an error summary when there are no errors > expected html 1`] = `<div />`;

exports[`ErrorSummary > should render an error summary with 1 error > expected html 1`] = `
<div>
  <section
    class="my-5 border-4 border-red-600 p-4"
    role="alert"
    tabindex="-1"
  >
    <h2
      class="font-lato text-lg font-semibold"
    >
      {"key":"gcweb:error-summary.header","options":{"count":1}}
    </h2>
    <ol
      class="mt-1.5 list-decimal space-y-2 pl-7"
    >
      <li>
        <a
          class="text-red-700 underline hover:decoration-2 focus:decoration-2"
          data-testid="anchor-link"
          href="#input-test"
        >
          Test error 1
        </a>
      </li>
    </ol>
  </section>
</div>
`;

exports[`ErrorSummary > should render an error summary with 2 errors > expected html 1`] = `
<div>
  <section
    class="my-5 border-4 border-red-600 p-4"
    role="alert"
    tabindex="-1"
  >
    <h2
      class="font-lato text-lg font-semibold"
    >
      {"key":"gcweb:error-summary.header","options":{"count":2}}
    </h2>
    <ol
      class="mt-1.5 list-decimal space-y-2 pl-7"
    >
      <li>
        <a
          class="text-red-700 underline hover:decoration-2 focus:decoration-2"
          data-testid="anchor-link"
          href="#input-test"
        >
          Test error 1
        </a>
      </li>
      <li>
        <a
          class="text-red-700 underline hover:decoration-2 focus:decoration-2"
          data-testid="anchor-link"
          href="#input-test"
        >
          Test error 2
        </a>
      </li>
    </ol>
  </section>
</div>
`;

exports[`FetcherErrorSummary > displays errors from fetcher data > expected html 1`] = `
<div>
  <div>
    <section
      class="my-5 border-4 border-red-600 p-4"
      role="alert"
      tabindex="-1"
    >
      <h2
        class="font-lato text-lg font-semibold"
      >
        {"key":"gcweb:error-summary.header","options":{"count":1}}
      </h2>
      <ol
        class="mt-1.5 list-decimal space-y-2 pl-7"
      >
        <li>
          <a
            class="text-red-700 underline hover:decoration-2 focus:decoration-2"
            data-testid="anchor-link"
            href="#test-field"
          >
            Test error message
          </a>
        </li>
      </ol>
    </section>
    <div
      aria-errormessage="test-error"
      id="test-field"
    >
      Test Field
    </div>
    <div
      id="test-error"
    >
      Test error message
    </div>
  </div>
</div>
`;

exports[`FormErrorSummary > displays errors from form action data > expected html 1`] = `
<div>
  <div>
    <section
      class="my-5 border-4 border-red-600 p-4"
      role="alert"
      tabindex="-1"
    >
      <h2
        class="font-lato text-lg font-semibold"
      >
        {"key":"gcweb:error-summary.header","options":{"count":1}}
      </h2>
      <ol
        class="mt-1.5 list-decimal space-y-2 pl-7"
      >
        <li>
          <a
            class="text-red-700 underline hover:decoration-2 focus:decoration-2"
            data-testid="anchor-link"
            href="#test-field"
          >
            Test error message
          </a>
        </li>
      </ol>
    </section>
    <div
      aria-errormessage="test-error"
      id="test-field"
    >
      Test Field
    </div>
    <div
      id="test-error"
    >
      Test error message
    </div>
  </div>
</div>
`;
